load-test-framework/dotnet_src/Google.PubSub.Flic/Program.cs

```diff
@@ -0,0 +1,225 @@
+                    }
+                    switch (startRequest.ClientOptionsCase)
+                    {
+                        case StartRequest.ClientOptionsOneofCase.PublisherOptions:
+                            await Publish(startRequest, stats);
+                            break;
+                        case StartRequest.ClientOptionsOneofCase.SubscriberOptions:
+                        default: // no option is sent for subscribe
+                            await Subscribe(startRequest, stats);
+                            break;
+                    }
+                    await Task.Delay(TimeSpan.FromHours(6));
+                    return 0;
+                case NotParsed<Options> failure:
+                    return 1;
+                default:
+                    return 2;
+            }
+        }
+
+        static async Task Publish(StartRequest startRequest, Stats stats)
+        {
+            var topicName = new TopicName(startRequest.Project, startRequest.Topic);
+            var clientCreationSettings = new PublisherClient.ClientCreationSettings(
+                clientCount: (int)Math.Max(Environment.ProcessorCount * startRequest.CpuScaling, 1));
+            var settings = new PublisherClient.Settings
+            {
+                BatchingSettings = new BatchingSettings(
+                    elementCountThreshold: startRequest.PublisherOptions.BatchSize,
+                    byteCountThreshold: null,
+                    delayThreshold: startRequest.PublisherOptions.BatchDuration.ToTimeSpan())
+            };
+            var client = await PublisherClient.CreateAsync(topicName, clientCreationSettings,  settings);
+            int msgsSent = 0;
+            double rate = startRequest.PublisherOptions.Rate;
+            var startTime = DateTime.UtcNow;
+            var endTime = startTime + startRequest.TestDuration.ToTimeSpan();
+            var msg = new byte[Math.Max(12, startRequest.PublisherOptions.MessageSize)];
+            var rnd = new Random(1);
+            rnd.NextBytes(msg);
+            var msgsInFlight = 0;
+            const int MaxQueued = 1_000_000;
+            while (true)
+            {
+                while (true)
+                {
+                    var now = DateTime.UtcNow;
+                    if (now >= endTime)
+                    {
+                        stats.End(now);
+                        return;
+                    }
+                    var expectedMsgsSent = (now - startTime).TotalSeconds * rate;
+                    if ((rate == 0.0 || msgsSent < expectedMsgsSent) && Interlocked.Add(ref msgsInFlight, 0) < MaxQueued)
+                    {
+                        break;
+                    }
+                    await Task.Delay(TimeSpan.FromMilliseconds(1));
+                }
+                var msgId = msgsSent;
+                var publishTime = DateTime.UtcNow;
+                Array.Copy(BitConverter.GetBytes(msgId), 0, msg, 0, 4);
```

Comment:
Do not use bit bashing.  The api takes attributes, store message id and publish time in those fields using the api https://googleapis.github.io/google-cloud-dotnet/docs/Google.Cloud.PubSub.V1/api/Google.Cloud.PubSub.V1.PubsubMessage.html

This is what the java client does https://github.com/GoogleCloudPlatform/pubsub/blob/master/load-test-framework/src/main/java/com/google/pubsub/clients/gcloud/CPSPublisherTask.java